### PR TITLE
Provide a clear error when received message args are mutated.

### LINKED ISSUE #865: spies don't tolerate changing arguments
```ruby
describe 'rspec-mocks' do
  it 'passes with a double' do
    d = double('d').as_null_object
    expect(d).to receive(:m).with(hash_including('offset' => 1))
    args = {}
    args['offset'] = 1
    d.m args
    args['offset'] = 2
    d.m args
  end

  it 'passes with a spy' do
    d = spy('d')
    args = {}
    args['offset'] = 1
    d.m args
    args['offset'] = 2   # it passes if you comment
    d.m args             # out these two lines
    expect(d).to have_received(:m).with(hash_including('offset' => 1))
  end
end
```

### PR DESCRIPTION
**Pull request by myronmarston**
...before a `have_received(...).with(...)` expectation.

Fixes #865.

### CODE REVIEW

File: lib/rspec/mocks/proxy.rb [hunk 1]
```diff
@@ -149,7 +156,35 @@ def has_negative_expectation?(message)
...
-        @messages_received << [message, args, block]
+        @messages_received << [message, ReceivedArgList.new(args), block]
+      end
+
+      class ReceivedArgList
+        attr_reader :args
+
+        def initialize(args)
+          @args = args
+          @original_hashes = args.map { |arg| [arg, hash_of(arg)] }
```
Comment (jferris) [inline lib/rspec/mocks/proxy.rb]: `Array#hash` will return a `hash` based on its components, so I think you can get away with just doing `@original_hash = @args.hash` and then comparing that later on.

Comment (myronmarston) [inline lib/rspec/mocks/proxy.rb]: You're right.  Thanks for pointing that out!

**Comment by JonRowe** [comment]
Nice work :)

File: lib/rspec/mocks/error_generator.rb [hunk 2]
```diff
@@ -15,6 +15,10 @@ module Mocks
     # invoked.
     MockExpectationAlreadyInvokedError = Class.new(Exception)
 
+    # Raised for situations that RSpec cannot support due to external mutations
```
Comment (waterlink) [inline lib/rspec/mocks/error_generator.rb]: "due to external mutations made externally" - sounds a bit odd to me: "external" part is doubled.

**Pull request merged**